### PR TITLE
RavenDB-4866

### PR DESCRIPTION
**Pull request by redknightlois**
RavenDB-4866: Performance degradation is being caused by memory fragmentation.

The current ByteStringContext change will try to diminish the fragmentation keeping track of a global pool. While potentially complicated to fix the issue for good, this is an initial approximation (good enough) to deal with the issue. More strict long term performance profiling should be performed to ensure that this strategy can lower fragmentation of the unmanaged heap to the point we need it.

### CODE REVIEW

File: src/Sparrow/ByteString.cs [hunk 1]
```diff
@@ -605,8 +661,22 @@ private ByteString Create(void* ptr, int length, int size, ByteStringType type =
...
+
+            SegmentInformation segment;
+            if (length > MaxAllocationBlockSizeInBytes)
+            {
+                var memory = new UnmanagedGlobalSegment(length);
+                byte* start = (byte*)memory.Segment.ToPointer();
+                byte* end = start + memory.Size;
+                                
+                segment = new SegmentInformation { Memory = memory, Start = start, Current = start, End = end, CanDispose = true };
+                _wholeSegments.Add(segment);
```
Comment (ayende) [inline src/Sparrow/ByteString.cs:674]: Aren't you adding the segment to the `_wholeSegments` collection twice? Once here and once after the if?

**Comment by ayende** [comment]
Failing test:

```
   FastTests.Sparrow.ByteString.AllocateAndReleaseShouldReuse [FAIL]
      Assert.NotInRange() Failure
      Range:  (2392899113824 - 2392899179360)
      Actual: 2392899179360
      Stack Trace:
        C:\Work\ravendb-4.0\test\FastTests\Sparrow\ByteString.cs(88,0): at FastTests.Sparrow.ByteString.AllocateAndReleaseShouldReuse()
```

File: src/Sparrow/ByteString.cs [hunk 2]
```diff
@@ -725,10 +796,19 @@ public void Release(ref ByteString value)
 
         private SegmentInformation AllocateSegment(int size)
```
Comment (ayende) [inline src/Sparrow/ByteString.cs:797]: There is some code duplication between `AllocateSegment` and `AllocateExternalSegment` that should probably be merged

File: src/Sparrow/ByteString.cs [hunk 3]
```diff
@@ -737,10 +817,19 @@ private SegmentInformation AllocateSegment(int size)
 
         private void AllocateExternalSegment(int size)
```
Comment (ayende) [inline src/Sparrow/ByteString.cs:818]: Are there any promises regarding the `size` ? Is it always power of 2 (or rather, power of 64K) ?

File: src/Sparrow/ObjectPool.cs [hunk 4]
```diff
@@ -274,4 +275,280 @@ private void Validate(object obj)
...
+    /// 
+    /// 2) it is implied that if object was obtained from a pool, the caller will return it back in
+    ///    a relatively short time. Keeping checked out objects for long durations is ok, but 
+    ///    reduces usefulness of pooling. Just new up your own.
+    /// 
+    /// Not returning objects to the pool in not detrimental to the pool's work, but is a bad practice. 
+    /// Rationale: 
+    ///    If there is no intent for reusing the object, do not use pool - just use "new". 
+    /// </summary>
+    public class ObjectPool<T, TFactoryParam> where T : class
```
Comment (ayende) [inline src/Sparrow/ObjectPool.cs:298]: Do we really need to duplicate all this code? Can we have a single `ObjectPool<T, TFactoryParam>` and then a inherited class `ObjectPool<T>` that just specialize it?

File: src/Sparrow/ByteString.cs (hunk 3 above)
Comment (redknightlois) [inline src/Sparrow/ByteString.cs:818]: No, there is no promise of any particular size. You allocate whatever you want. The only promise is that it cannot be lower than 64Kb.

File: src/Sparrow/ObjectPool.cs (hunk 4 above)
Comment (redknightlois) [inline src/Sparrow/ObjectPool.cs:298]: Probably, wasnt my objective though. I can refactor that sure.

File: src/Sparrow/ByteString.cs (hunk 1 above)
Comment (redknightlois) [inline src/Sparrow/ByteString.cs:674]: Adding only happens at the AllocateSegment. This is basically an stripped down version of it.

Comment (redknightlois) [inline src/Sparrow/ByteString.cs:674]: In fact, I just realized we can do with far less code. I will change that tomorrow morning.

Comment (ayende) [inline src/Sparrow/ByteString.cs:674]: See:
https://github.com/Corvalius/ravendb/blob/cf1462c1ea3408bfee6fb26abc19e1eac881d715/src/Sparrow/ByteString.cs#L674
https://github.com/Corvalius/ravendb/blob/cf1462c1ea3408bfee6fb26abc19e1eac881d715/src/Sparrow/ByteString.cs#L683

**Comment by ayende** [comment]
The reason the test fails is because is assumes certain sizes for the segments, but because we pooling the memory, we are actually seeing different sizes, so it breaks.
It will always pass when run on its own, but sometimes fail when run in all the tests.

**Comment by ayende** [comment]
Here is a consistent failing test below.
I skipped the failing tests for now, please fix them as part of the rest of the work.

```

        [Fact]
        public void ConstructionInsideWholeSegmentWithHistory()
        {
            using (var context = new ByteStringContext(ByteStringContext.MinBlockSizeInBytes))
            {
                for (int i = 0; i < 10; i++)
                {
                    context.Allocate(ByteStringContext.MinBlockSizeInBytes * 2);
                }
            }
            using (new ByteStringContext(ByteStringContext.MinBlockSizeInBytes))
            using (var context = new ByteStringContext(ByteStringContext.MinBlockSizeInBytes))
            {
                var byteStringInFirstSegment = context.Allocate((ByteStringContext.MinBlockSizeInBytes / 2) - sizeof(ByteStringStorage));
                var byteStringWholeSegment = context.Allocate((ByteStringContext.MinBlockSizeInBytes / 2) - sizeof(ByteStringStorage));
                var byteStringNextSegment = context.Allocate(1);

                long startLocation = (long)byteStringInFirstSegment._pointer;
                Assert.InRange((long)byteStringWholeSegment._pointer, startLocation, startLocation + ByteStringContext.MinBlockSizeInBytes);
                Assert.NotInRange((long)byteStringNextSegment._pointer, startLocation, startLocation + ByteStringContext.MinBlockSizeInBytes);
            }
        }
```

**Pull request merged**